### PR TITLE
fix(notification-center): api calls on first render

### PR DESCRIPTION
**Pull request by sashasushko**
### What kind of change does this PR introduce?

Bug fix

### Why was this change needed?

React `useEffect` can't detect changes in out of scope variables like `api` (look the Note block in [`useEffect` optimizations docs](https://reactjs.org/docs/hooks-effect.html#tip-optimizing-performance-by-skipping-effects)).

[Here is demo](https://codepen.io/sashasushko/pen/XWYrOdJ) what I mean.

That's why on first render the `initialized` prop of `NovuContext.Provider` will always be `false`. Which to infinity loading bug. Because of this, the user sees [infinity loading instead of the content](https://github.com/novuhq/novu/blob/main/packages/notification-center/src/components/notification-center/components/layout/Layout.tsx#L29).

### CODE REVIEW

File: packages/notification-center/src/stories/notification-center.stories.tsx [hunk 1]
```diff
@@ -10,6 +10,22 @@ export default {
   },
 } as ComponentMeta<typeof NotificationCenter>;
 
-const templateElement: ComponentStory<typeof NotificationCenter> = (args) => <NotificationCenter {...args} />;
+const Component = (...props) => {
```
Comment (sashasushko) [inline packages/notification-center/src/stories/notification-center.stories.tsx:13]: This is for testing purposes only. I will clean up it before merging

File: packages/notification-center/src/stories/notification-center.stories.tsx [hunk 2]
```diff
@@ -10,6 +10,22 @@ export default {
...
+  const [subscriberId, setSubscriberId] = useState(undefined);
+
+  useEffect(() => {
+    setTimeout(() => {
+      setSubscriberId('a9b752b2-1cae-48ce-a1e0-9d07c3650af0');
+    }, 3000);
+  }, []);
+
+  return (
+    <NovuProvider subscriberId={subscriberId} applicationIdentifier="...">
```
Comment (sashasushko) [inline packages/notification-center/src/stories/notification-center.stories.tsx:23]: You should provide real `applicationIdentifier` to see changes in Storybook

File: packages/notification-center/src/components/novu-provider/NovuProvider.tsx [hunk 3]
```diff
@@ -73,47 +67,39 @@ export function NovuProvider(props: INovuProviderProps) {
...
-  async function initSession(payload: { clientId: string; data: { subscriberId: string }; subscriberHash: string }) {
+  const initSession = useCallback(async () => {
     if ('parentIFrame' in window) {
       // eslint-disable-next-line @typescript-eslint/no-explicit-any
       (window as any).parentIFrame.autoResize(true);
     }
 
-    const response = await apiService.initializeSession(payload.clientId, payload.data.subscriberId, subscriberHash);
-
-    api.setAuthorizationToken(response.token);
```
Comment (sashasushko) [inline packages/notification-center/src/components/novu-provider/NovuProvider.tsx:106]: This method will call inside `applyToken`, so there isn't necessary to call it separately

File: packages/notification-center/src/components/novu-provider/NovuProvider.tsx [hunk 4]
```diff
@@ -57,7 +51,7 @@ export function NovuProvider(props: INovuProviderProps) {
     >
       <ApiContext.Provider value={{ api }}>
         <AuthProvider>
-          <SessionInitialization applicationIdentifier={props.applicationIdentifier} subscriberId={props.subscriberId}>
+          <SessionInitialization onInit={setIsSessionInitialized}>
```
Comment (sashasushko) [inline packages/notification-center/src/components/novu-provider/NovuProvider.tsx:54]: This is hotfix. But the right way is to create provider's chain:
```
<ApiProvider>  <- creates api instance 
  <AuthProvider>  <- prepares "applyToken", "setUser"
    <SessionInitializationProvider>  <- calls "initSession" and passes "initialized" flag to children instead NovuContext.Provider
```

File: packages/notification-center/src/components/notification-center/components/notification-item/NotificationListItem.tsx [hunk 5]
```diff
@@ -4,9 +4,8 @@ import { formatDistanceToNow } from 'date-fns';
 import { IMessage, ButtonTypeEnum, IMessageAction, MessageActionStatusEnum } from '@novu/shared';
 import { DotsHorizontal } from '../../../../shared/icons';
 import { INovuTheme } from '../../../../store';
-import { useNovuTheme, useNotificationCenter } from '../../../../hooks';
+import { useNovuTheme, useNotificationCenter, useTranslations } from '../../../../hooks';
 import { ActionContainer } from './ActionContainer';
-import { useTranslations } from 'packages/notification-center/src/hooks/use-translations';
```
[the remark this answers is not in the document]
Comment (scopsy) [reply packages/notification-center/src/components/notification-center/components/notification-item/NotificationListItem.tsx:9]: Yep! looks like, thanks!

**Comment by scopsy** [comment]
@djabarovgeorge could you please take a look at this, encountered this while working on the novu demo when initializing a page with an already open notification center component without a popover.

**Comment by djabarovgeorge** [comment]
Hey, @sashasushko 
Great job for the initiative to resolve this issue.
Did you have a chance to test the updated components?

**Comment by sashasushko** [comment]
>  Did you have a chance to test the updated components?

Only locally in your dev environment. So, if you have some sandbox app similar to real one, it will be a great idea to test changes there.

**Comment by djabarovgeorge** [comment]
@sashasushko Not a problem I will look into it, Is there a way to reproduce the infinity loading bug? Also, I can see the tests are failing have you had a chance to check them?

**Comment by sashasushko** [comment]
> Is there a way to reproduce the infinity loading bug?

Just use instruction from docs https://docs.novu.co/notification-center/getting-started/#react-component with some real `APP_ID_FROM_ADMIN_PANEL` and `USER_ID`

**Comment by sashasushko** [comment]
> the tests are failing have you had a chance to check them?

🆗

**Comment by sashasushko** [comment]
@djabarovgeorge Could you help me with running the tests locally? I have "infinity executing" on my MacBook m1 and no results
Снимок экрана 2022-11-03 в 21 22 12

**Comment by LetItRock** [comment]
@sashasushko hey, if you want to run the e2e tests for the Widget locally you should:
- in the three folders: API, WS, and Widget you should run `npm run start:test`
- then again in the Widget folder `npm run cypress:open` to run e2e tests

**Comment by LetItRock** [comment]
I also see a bug in the `get-notifications-feed.usecase.ts:49`. It was probably a merge issue that wasn't correctly resolved @jainpawan21 @djabarovgeorge 
it should be like this: `const totalCount = await this.messageRepository.getUnseenCount`
so this change will help your local tests pass

**Comment by djabarovgeorge** [comment]
Hey @LetItRock, Maybe I didn't understand you correctly, but I am not sure about your suggestion regarding the change, because it is related to the total notification number and not to the total unseen notification number.

**Comment by LetItRock** [comment]
@djabarovgeorge hey, you are right, I didn't check the context. Sorry for bothering.

**Comment by djabarovgeorge** [comment]
@LetItRock No bother was done! :) 

@sashasushko Sorry for the confusion looks like I was a bit out of context, Dima pointed me to the Demo you were working on where this issue occurred, I tested your fix and it resolves it well, so thanks for the PR you raised! it helped a lot.

**Comment by sashasushko** [comment]
@djabarovgeorge you're welcome!

**Pull request closed without merging**